rfcs/2020-08-31-3641-mongo-metrics.md

```diff
@@ -0,0 +1,801 @@
+# RFC 3603 - 2020-08-27 - Collecting metrics from MongoDB
+
+This RFC is to introduce a new metrics source to consume metrics from MongoDB database servers. The high level plan is to implement one source that collects metrics from MongoDB server instances.
+
+Background reading on MongoDB monitoring:
+
+- https://docs.mongodb.com/manual/administration/monitoring/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for MongoDB server metrics.
+
+This RFC will not cover:
+
+- Other databases.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their MongoDB databases are performing.
+
+## Internal Proposal
+
+Build a single source called `mongodb_metrics` (name to be confirmed) to collect MongoDB metrics. We support all non-EOL'ed MongoDB versions.
+
+The recommended implementation is to use the Rust MongoDB client to connect the target database server by the address specified in configuration.
+
+Proposed client:
+
+- https://github.com/mongodb/mongo-rust-driver
+- https://crates.io/crates/mongodb
+
+We'd call the following on the server from the client upon connection:
+
+- [`serverStatus`](https://docs.mongodb.com/manual/reference/command/serverStatus/)
+- [`printShardingStatus`](https://docs.mongodb.com/manual/reference/method/db.printShardingStatus/)
+- [`shardConnPoolStats`](https://docs.mongodb.com/manual/reference/command/shardConnPoolStats/)
+- [`dbStats`](https://docs.mongodb.com/manual/reference/command/dbStats/) on each DB
+- [`collStats`](https://docs.mongodb.com/manual/reference/command/collStats/) per collection
+- [`connPoolStats`](https://docs.mongodb.com/manual/reference/command/connPoolStats/) per pool
+- [`Top`](https://docs.mongodb.com/manual/reference/command/top/)
+- [`indexStats`](https://docs.mongodb.com/manual/reference/operator/aggregation/indexStats/#pipe._S_indexStats)
+- [`replSetGetStatus`](https://docs.mongodb.com/manual/reference/command/replSetGetStatus/)
+
+This should result in the following metrics being collected:
+
+- `mongodb_up` -> Used as an uptime metric with 1 for successful collection and 0 for failed collection (gauge)
+- `mongodb_asserts_total` (counter)
+- `mongodb_connections` (gauge)
+- `mongodb_connections_metrics_created_total` (counter)
+- `mongodb_extra_info_heap_usage_bytes` (gauge)
+- `mongodb_extra_info_page_faults` (gauge)
+- `mongodb_instance_local_time` (gauge)
+- `mongodb_instance_uptime_estimate_seconds_total` (counter)
```

Comment:
It's still not clear to me what the difference is between this one and `mongodb_instance_uptime_seconds_total`; even after looking at https://docs.mongodb.com/manual/reference/command/serverStatus/ . Are you aware? The number seems to be the same when I ran a local mongodb docker container.

Reply:
I actually can't work out the difference. Maybe it's a replaced/deprecated metric?

Reply:
It is definitely unclear to me :smile: but I'm fine with reporting both given that `serverStatus` does. Maybe they differ in certain circumstances.